Added IaC example page documentation (#1610)
liqotech/liqo: docs/examples/quick-start-with-terraform.md

> # Quick Start with Terraform

Reviewer: ```suggestion
# Using Terraform to play with Liqo
```

> This tutorial aims at presenting how to set up an environment with Liqo installed via Terraform and explore the Liqo capabilities through its Provider and Resources.

Reviewer: ```suggestion
Terraform is a widely used Infrastructure as Code (IaC) tool that allows engineers to define their software infrastructure in code.
This tutorial presents how to leverage Terraform to set up two Kubernetes clusters peered with Liqo, and explore the Liqo capabilities through its Provider and Resources.
```

## Provision the infrastructure

Before starting the tutorials below, you should ensure the following software is installed on your system:

- [**Docker**](https://docker.io), the container runtime.
- [**Kubectl**](https://kubernetes.io/docs/tasks/tools/install-kubectl/), the command-line tool for Kubernetes.
- [**Terraform**](https://developer.hashicorp.com/terraform/tutorials/aws-get-started/install-cli), the command-line tool for Terraform.

> Then, let's open a terminal on your machine and launch the following script, which creates the complete infrastructure with all playground already set up.

Reviewer: ```suggestion
Then, let's open a terminal on your machine and launch the following script, which creates the infrastructure defined in this example (i.e., two `kind` clusters, peered with Liqo), with all playground already set up.
This tutorial will present a detailed description about how this result is achieved, by analyzing the most notable parts of the Terraview infrastructure definition file that refer to Liqo.
```

{{ env.config.html_context.generate_clone_example_tf('quick-start-terraform') }}

## Analyze the infrastructure and code

> With the previous command you created two KinD clusters, peered with each other. Furthermore, you offloaded a namespace in Liqo virtual cluster.\

Reviewer: ```suggestion
The previous command created two KinD clusters, peered with each other. Furthermore, you offloaded a namespace in Liqo virtual cluster.\
```
Be more specific. How are resources offloaded? A offloads to B? Viceversa? Bi-directional?

> The choice of use of KinD was made because of there are no configuration requirements such as account creation or resources costs, therefore all the presented functionality works also on Cloud provider clusters.

Reviewer: ```suggestion
KinD was chosen because of there are no configuration requirements such as account creation or resource costs, therefore all the presented functionalities works also on other clusters, e.g., the ones operated by public cloud providers.
```

### KinD resource

> The first executed step by Terraform is the the creation of the KinD clusters: the resource in charge of building them is the `kind_cluster` resource of the provider [tehcyx](https://registry.terraform.io/providers/tehcyx/kind/latest/docs) which starting from configuration parameters (such as cluster_name, service_subnet/pod_subnet) generates the clusters and related config files needed by other providers to set up the infrastructure.

Reviewer: ```suggestion
The first executed step by Terraform is the the creation of the two KinD clusters: the resource in charge of building them is the `kind_cluster` resource of the provider [tehcyx](https://registry.terraform.io/providers/tehcyx/kind/latest/docs) that, starting from configuration parameters (such as cluster_name, service_subnet/pod_subnet), generates the clusters and related config files needed by other providers to set up the infrastructure.
```

You can inspect the deployed clusters by typing on your workstation:

```bash
kind get clusters
```

You should see a couple of entries:

```text
milan
rome
```

This means that two KinD clusters are deployed and running on your host.

Then, you can simply inspect the status of the clusters.
To do so, you can export the `KUBECONFIG` variable to specify the identity file for _kubectl_ and _liqoctl_, and then contact the cluster.

By default, the kubeconfigs of the two clusters are stored in the current directory (`./rome-config`, `./milan-config`).
You can export the appropriate environment variables leveraged for the rest of the tutorial (i.e., `KUBECONFIG` and `KUBECONFIG_MILAN`), and referring to their location, through the following:

```bash
export KUBECONFIG="$PWD/rome-config"
export KUBECONFIG_MILAN="$PWD/milan-config"
```

```{admonition} Note
We suggest exporting the kubeconfig of the first cluster as default (i.e., `KUBECONFIG`), since it will be the entry point of the virtual cluster and you will mainly interact with it.
```

### Install Liqo with Helm resource

> After creating the clusters KinD Terraform will install Liqo using the resource `helm_release` of the provider [Helm](https://registry.terraform.io/providers/hashicorp/helm/latest/docs) once correctly configured with the cluster config files.

Reviewer: ```suggestion
After creating the two KinD clusters, Terraform will install Liqo using the `helm_release` resource of the [Helm provider](https://registry.terraform.io/providers/hashicorp/helm/latest/docs), once correctly configured with the cluster config files.
```
I do not understand exactly. With "once correctly configured with the cluster config files", do you mean:
- after having correctly configured the cluster config files
- (anything else?)

Author: We mean the correct configuration of the provider through config file or parameters.

> Once installation is completed the Liqo pods should be up and running on both clusters:

Reviewer: ```suggestion
Once installation is completed, you should see the Liqo pods up and running on both clusters:
```

```bash
kubectl get pods -n liqo
```

```text
NAME                                       READY   STATUS    RESTARTS   AGE
liqo-auth-74c795d84c-x2p6h                 1/1     Running   0          2m8s
liqo-controller-manager-6c688c777f-4lv9d   1/1     Running   0          2m8s
liqo-crd-replicator-6c64df5457-bq4tv       1/1     Running   0          2m8s
liqo-discovery-546fd594dc-kg6db            1/1     Running   0          2m8s
liqo-gateway-78cf7bb86b-pkdpt              1/1     Running   0          2m8s
liqo-metric-agent-5667b979c7-snmdg         1/1     Running   0          2m8s
liqo-network-manager-5b5cdcfcf7-scvd9      1/1     Running   0          2m8s
liqo-proxy-6674dd7bbd-kr2ls                1/1     Running   0          2m8s
liqo-route-7wsrx                           1/1     Running   0          2m8s
liqo-route-sz75m                           1/1     Running   0          2m8s
liqo-webhook-7bb6dbf88d-2hvwx              1/1     Running   0          2m8s
```

### Liqo generate resource

> Once Liqo is installed in remote cluster, Terraform will generate the authentication parameters to peer the remote cluster with the local one.\

Reviewer: ```suggestion
Once Liqo is installed in both clusters, Terraform will generate the authentication parameters to peer the remote cluster with the local one.\
```

Author: Following dependencies wrote in Terraform example file the generate resource will start after creation of remote cluster and not after the creation of both.

Is this not correct in Liqo context? Has generate resource to start after creation of both cluster?

Reviewer: @AleDecre I don't understand. Liqo has to be installed in **both** clusters. Once this isdone, we can create the proper resources to set-up the peering.
So, the original sentence that says that Liqo is installed in the remote cluster looks misleading to me, as it must be installed in both.

Author: What we meant is that Generate Resource has no dependency on the liqo installation on the local cluster, so when you run the Terraform script, at the moment, it just waits for the installation to complete on the remote cluster.
Therefore, if liqo is already installed on the remote cluster but not yet on the local one, Generate Resource will still run.
Is this correct or is it more correct to wait for the local cluster to be ready as well before execute the Generate Resource? In that case we have to modify the dependencies in the main.tf.

Reviewer: @AleDecre Got it.
I would suggest to rephrase this way: 
`Once the Liqo installation in the remote cluster is completed, Terraform will generate`....

> The resource in charge of doing this is `liqo_generate` of `liqo` provider configured with config file or parameters of remote cluster:

Reviewer: ```suggestion
This is achieved with the `liqo_generate` resource of the `liqo` Terraform provider, configured with either the config file or the full list of parameters of the remote cluster:
```
(not sure I got the message right; please check my rephrasing)

```
provider "liqo" {
  alias = "milan"
  kubernetes = {
    config_path = kind_cluster.milan.kubeconfig_path
  }
}

resource "liqo_generate" "generate" {

  depends_on = [
    helm_release.install_liqo_milan
  ]

  provider = liqo.milan

}
```

### Liqo peering resource

> Once `generate_resource` is created, Terraform will continue peering procedure with the output parameters of the previous resource.\

Reviewer: ```suggestion
Once the `generate_resource` is created, Terraform will continue the peering procedure leveraging the output parameters of the previous resource.\
```

> The resource in charge of doing this is `liqo_peering` of `liqo` provider configured with config file or parameters of local cluster:

Reviewer: ```suggestion
This is achieved with the `liqo_peering` resource of the `liqo` Terraform provider, configured with either the config file or the full list of parameters of the local cluster:
```

  cluster_name    = liqo_generate.generate.cluster_name
  cluster_authurl = liqo_generate.generate.auth_ep
  cluster_token   = liqo_generate.generate.local_token

}
```

You can check the peering status by running:

```bash
kubectl get foreignclusters
```

The output should look like the following, indicating that the cross-cluster network tunnel has been established, and an outgoing peering is currently active (i.e., the _Rome_ cluster can offload workloads to the _Milan_ one, but not vice versa):

```text
NAME    TYPE        OUTGOING PEERING   INCOMING PEERING   NETWORKING    AUTHENTICATION   AGE
milan   OutOfBand   Established        None               Established   Established      12s
```

At the same time, you should see a virtual node (`liqo-milan`) in addition to your physical nodes:

```bash
kubectl get nodes
```

```text
NAME                 STATUS   ROLES                  AGE     VERSION
liqo-milan           Ready    agent                  14s     v1.23.6
rome-control-plane   Ready    control-plane,master   7m56s   v1.23.6
rome-worker          Ready    <none>                 7m25s   v1.23.6
```

### Liqo offload resource

> If you want to deploy an application that is scheduled onto Liqo virtual nodes, you should first create a namespace where your pod will be started.\

Reviewer: ```suggestion
If you want to deploy an application that is scheduled on a Liqo virtual node (hence, it is offloaded on a remote cluster), you should first create a namespace where your pod will be started.\
```

> This can be achieved with `kubernetes_namespace` of [kubernetes](https://registry.terraform.io/providers/hashicorp/kubernetes/latest/docs) provider configured with config file or parameters of local cluster.

Reviewer: ```suggestion
This can be achieved with `kubernetes_namespace` of [kubernetes](https://registry.terraform.io/providers/hashicorp/kubernetes/latest/docs) Terraform provider, configured with either the config file or the full lit of parameters of the local cluster.
```

Then tell Liqo to make this namespace eligible for the pod offloading.\
The resource in charge of doing this is `liqo_offload` of the same `liqo` provider of `liqo_peering` resource:

```
resource "liqo_offload" "offload" {

  depends_on = [
    helm_release.install_liqo_rome,
    kubernetes_namespace.namespace
  ]

  provider = liqo.rome

  namespace = "liqo-demo"

}
```

```{admonition} Note
Liqo virtual nodes have a [taint](https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/) that prevents the pods from being scheduled on them.
The Liqo webhook will add the toleration for this taint to the pods created in the liqo-enabled namespaces.
```

Since no further configuration is provided, Liqo will add a suffix to the namespace name to make it unique on the remote cluster (see the dedicated [usage page](/usage/namespace-offloading.md) for additional information concerning namespace offloading configurations).

> To test the infrastructure just created through the deploy of an application you can follow the relative example section in [Quick Start page](/examples/quick-start.md).

Reviewer: ```suggestion
You can now test the infrastructure you have just created by deploying an application.
For this, you can follow the proper example section in the [Quick Start page](/examples/quick-start.md).
```

> # Using Terraform to play with Liqo

Reviewer 2: Alternative proposal: *Provision with Terraform*

To me, mentioning Liqo in the title is redundant (and makes it longer than the others), while *play* conveys a somewhat limited message. @frisso, @aleoli WDYT?

Reviewer: LGTM

Reviewer 3: Yes, I completely agree with you

Terraform is a widely used Infrastructure as Code (IaC) tool that allows engineers to define their software infrastructure in code.\

> This tutorial aims at presenting how to set up an environment with Liqo installed via Terraform and explore the Liqo capabilities through its Provider and Resources.

Reviewer 2: I would personally suggest to better emphasize what a reader is expected to learn reading this tutorial.

## Provision the infrastructure

Before starting the tutorials below, you should ensure the following software is installed on your system:

- [**Docker**](https://docker.io), the container runtime.
- [**Kubectl**](https://kubernetes.io/docs/tasks/tools/install-kubectl/), the command-line tool for Kubernetes.
- [**Terraform**](https://developer.hashicorp.com/terraform/tutorials/aws-get-started/install-cli), the command-line tool for Terraform.

Then, let's open a terminal on your machine and launch the following script, which creates the infrastructure used in this example (i.e., two KinD clusters, peered with Liqo), with all playground already set up.\

> This tutorial will present a detailed description about how this result is achieved, by analyzing the most notable parts of the Terraview infrastructure definition file that refer to Liqo.

Author: Typo, Terraform instead of Terraview

{{ env.config.html_context.generate_clone_example_tf('quick-start-terraform') }}

## Analyze the infrastructure and code

> The previous command created two KinD clusters, peered with each other. Furthermore, you offloaded a namespace in Liqo virtual cluster.\

Reviewer 2: I would mention (and possibly link the file which contains the terraform configuration), for the sake of clarity. Otherwise it is unclear how the above commands provision the playground. I would also better clarify the setup (e.g., the peering is unidirectional, while from your sentence it seems bi-directional, and you do not mention that liqo is installed)

> In more details the offloaded namespace created in local cluster will leverage also on virtual one which isa resource abstraction of remote cluster.

Reviewer 2: It is unclear to me the meaning of this sentence.

> KinD was chosen because of there are no configuration requirements such as account creation or resource costs, therefore all the presented functionalities works also on other clusters, e.g., the ones operated by public cloud providers.

Reviewer 2: ```suggestion
This example is provisioned on KinD, since it requires no particular configurations (e.g., concerning accounts), and does not lead to resource costs. 
Yet, all the presented functionalities work also on other clusters, e.g., the ones operated by public cloud providers.
```

```text
NAME    TYPE        OUTGOING PEERING   INCOMING PEERING   NETWORKING    AUTHENTICATION   AGE
milan   OutOfBand   Established        None               Established   Established      12s
```

At the same time, you should see a virtual node (`liqo-milan`) in addition to your physical nodes:

```bash
kubectl get nodes
```

```text
NAME                 STATUS   ROLES                  AGE     VERSION
liqo-milan           Ready    agent                  14s     v1.23.6
rome-control-plane   Ready    control-plane,master   7m56s   v1.23.6
rome-worker          Ready    <none>                 7m25s   v1.23.6
```

### Liqo offload resource

If you want to deploy an application that is scheduled on a Liqo virtual node (hence, it is offloaded on a remote cluster), you should first create a namespace where your pod will be started.\
This can be achieved with `kubernetes_namespace` of [kubernetes](https://registry.terraform.io/providers/hashicorp/kubernetes/latest/docs) Terraform provider, configured with either the config file or the full lit of parameters of the local cluster.

Then tell Liqo to make this namespace eligible for the pod offloading.\

> The resource in charge of doing this is `liqo_offload` of the same `liqo provider` of `liqo_peering` resource:

Author: We wanted to say that in the example there are two liqo provider instantiation, one for remote cluster (generate resource) and one for local one (peering and offload resource).
Can we skip this note? Or have we to rephrase introducing and explaining this aspect?

> The resource in charge of doing this is `liqo_offload` of the same `liqo provider` of `liqo_peering` resource:

Reviewer 2: OK, then I would personally call that a provider instance. Although I'm not sure whether there is a better terminology in Terraform.

# Quick Start with Terraform

> This tutorial aims at presenting how to set up an environment with Liqo installed via Terraform and explore its most notable capabilities.

Reviewer: Not clear who "notable" refers to? Is it Terraform that has notable capabilities? Or Liqo?
Please clarify by rephrasing.

Author: We will try to clarify the aims by rephrasing in next commit.

You will create a _virtual cluster_ by peering two Kubernetes clusters and deploy a simple application on it.

## Provision the infrastructure

Before starting the tutorials below, you should ensure the following software is installed on your system:

- [**Docker**](https://docker.io), the container runtime.
- [**Kubectl**](https://kubernetes.io/docs/tasks/tools/install-kubectl/), the command-line tool for Kubernetes.
- [**Terraform**](https://developer.hashicorp.com/terraform/tutorials/aws-get-started/install-cli), the command-line tool for Terraform.

Then, let's open a terminal on your machine and launch the following script, which creates the complete infrastructure with all playground already set up.

> {{ env.config.html_context.generate_clone_example_tf('quick-start-terraform') }}

Reviewer: Not clear to me what you mean with "complete infrastructure with all playground already set up". More details (of, even better, a picture) can be helpful here. Otherwise I know that the script creates something, but I have no idea what.

## Analyze the infrastructure and code

> With this command two KinD clusters have been created, peered with each other and a namespace has been offloaded on which two pods and a service have been deployed.

Reviewer: ```suggestion
With the previous command you created two KinD clusters, peered with each other. Furthermore, you deployed two pods and a service in a namespace that has been offloaded.
```
Not clear to me where pods and services are created (on the first cluster? On the second?)

Author: Clarification about the pods and service scheduling in virtual cluster (and virtual nodes of course) will be present in next commit.

### KinD resource

> Let's start from the creation of the KinD clusters: the resource in charge of building them is the `kind_cluster` resource of the provider [tehcyx](https://registry.terraform.io/providers/tehcyx/kind/latest/docs) which starting from configuration parameters (such as cluster_name, service_subnet/pod_subnet) generates the clusters and related config files needed by other providers to set up the infrastructure.

Reviewer: Why are you saying "Let's start from the creation of the KinD cluster"? Are the cluster already up and running, or you need to create them now?
The message is not clear to me.

Author: Our intention was to analyze and comment step-by-step all the tasks performed by Terraform.

> You can inspect the deployed clusters by typing:

Reviewer: Where do you have to type this command? On your workstation? In a specific cluster?

Author: Updated with correct phrase.

```bash
kubectl get pods -n liqo
```

```text
NAME                                       READY   STATUS    RESTARTS   AGE
liqo-auth-74c795d84c-x2p6h                 1/1     Running   0          2m8s
liqo-controller-manager-6c688c777f-4lv9d   1/1     Running   0          2m8s
liqo-crd-replicator-6c64df5457-bq4tv       1/1     Running   0          2m8s
liqo-discovery-546fd594dc-kg6db            1/1     Running   0          2m8s
liqo-gateway-78cf7bb86b-pkdpt              1/1     Running   0          2m8s
liqo-metric-agent-5667b979c7-snmdg         1/1     Running   0          2m8s
liqo-network-manager-5b5cdcfcf7-scvd9      1/1     Running   0          2m8s
liqo-proxy-6674dd7bbd-kr2ls                1/1     Running   0          2m8s
liqo-route-7wsrx                           1/1     Running   0          2m8s
liqo-route-sz75m                           1/1     Running   0          2m8s
liqo-webhook-7bb6dbf88d-2hvwx              1/1     Running   0          2m8s
```

### Liqo generate resource

Once Liqo is installed in remote cluster, Terraform will generate the authentication parameters to peer the remote cluster with the local one.\

> The resource in charge of doing this is `liqo_generate` of [liqo](https://google.com) provider configured with config file or parameters of remote cluster:

Reviewer: The link looks wrong.

Author: We forget to remove this `placeholder`. The link has to be updated when Liqo Provider will be published on Terraform registry.

```text
NAME    TYPE        OUTGOING PEERING   INCOMING PEERING   NETWORKING    AUTHENTICATION   AGE
milan   OutOfBand   Established        None               Established   Established      12s
```

At the same time, you should see a virtual node (`liqo-milan`) in addition to your physical nodes:

```bash
kubectl get nodes
```

```text
NAME                 STATUS   ROLES                  AGE     VERSION
liqo-milan           Ready    agent                  14s     v1.23.6
rome-control-plane   Ready    control-plane,master   7m56s   v1.23.6
rome-worker          Ready    <none>                 7m25s   v1.23.6
```

### Liqo offload resource

If you want to deploy an application that is scheduled onto Liqo virtual nodes, you should first create a namespace where your pod will be started.\
This can be achieved with `kubernetes_namespace` of [kubernetes](https://registry.terraform.io/providers/hashicorp/kubernetes/latest/docs) provider configured with config file or parameters of local cluster.

Then tell Liqo to make this namespace eligible for the pod offloading.\

> The resource in charge of doing this is `liqo_offload` of the same [liqo](https://google.com) provider of `liqo_peering` resource:

Reviewer: The link looks broken.

Author: The Kubernetes or Liqo one?
If you mean Liqo one, it is our mistake as in the previuos comment. Sorry.
If you mean Kubernetes one, it seems to work.

```
resource "liqo_offload" "offload" {

  depends_on = [
    helm_release.install_liqo_rome,
    kubernetes_namespace.namespace
  ]

  provider = liqo.rome

  namespace = "liqo-demo"

}
```

```{admonition} Note

> The virtual nodes have a [taint](https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/) that prevents the pods from being scheduled on them.

Reviewer: ```suggestion
Liqo virtual nodes have a [taint](https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/) that prevents the pods from being scheduled on them.
```

Author: Committed correctly

> The Liqo webhook will add the toleration for this taint to the pods created in the liqo-enabled namespaces.

Reviewer: Add something like "so that those pods..."
(the sentence looks half the story to me: it says that it's adding a toleration, but it doesn't say to do what

Author: This section is removed following Marco Iorio's suggestions.

Reviewer: Personally, I still feel this comment is not clear. I would suggest to be more precise, hence avoiding that the reader has to be be 100% aligned with all the Liqo documentation in order to understand what's written here.

```

Since no further configuration is provided, Liqo will add a suffix to the namespace name to make it unique on the remote cluster (see the dedicated [usage page](/usage/namespace-offloading.md) for additional information concerning namespace offloading configurations).

### Deploy simple pod on offloaded namespace

At this point the basic setup is completed and next resources created by Terraform will deploy a standard Kubernetes application in a multi-cluster environment as you would do in a single cluster scenario.

> The deployed resources are t two pods running an `nginx` image and a Service exposing the pods to the cluster.

Reviewer: ```suggestion
The deployed resources are two pods running an `nginx` image and a Service exposing the pods to the cluster.
```

Author: This section is removed following Marco Iorio's suggestions.

```bash
kubectl get pod -n liqo-demo -o wide
```

And the output should look like this:

```text
NAME           READY   STATUS    RESTARTS   AGE   IP            NODE          NOMINATED NODE   READINESS GATES
nginx-local    1/1     Running   0          94s   10.200.1.10   rome-worker   <none>           <none>
nginx-remote   1/1     Running   0          94s   10.202.1.9    liqo-milan    <none>           <none>
```

#### Check the pod connectivity

Once both pods are correctly running, it is possible to check one of the abstractions introduced by Liqo.
Indeed, Liqo enables each pod to be transparently contacted by every other pod and physical node (according to the Kubernetes model), regardless of whether it is hosted by the _local_ or by the _remote_ cluster.

First, let's retrieve the IP address of the `nginx` pods:

```bash
LOCAL_POD_IP=$(kubectl get pod nginx-local -n liqo-demo --template={{.status.podIP}})
REMOTE_POD_IP=$(kubectl get pod nginx-remote -n liqo-demo --template={{.status.podIP}})
echo "Local Pod IP: ${LOCAL_POD_IP} - Remote Pod IP: ${REMOTE_POD_IP}"
```

> You can fire up a pod and run `curl` from inside the cluster:

Reviewer: ```suggestion
You can fire up a _client_ pod and run `curl` from inside the cluster:
```

Author: This section is removed following Marco Iorio's suggestions.

```bash
kubectl run --image=curlimages/curl curl -n default -it --rm --restart=Never -- curl ${LOCAL_POD_IP}
kubectl run --image=curlimages/curl curl -n default -it --rm --restart=Never -- curl ${REMOTE_POD_IP}
```

Both commands should lead to a successful outcome (i.e., return a demo web page), regardless of whether each pod is executed locally or remotely.

#### Expose the pods through a Service

The above configuration creates a Service which is designed to serve traffic to the previously deployed pods.

> This is a traditional [Kubernetes Service](https://kubernetes.io/docs/concepts/services-networking/service/) and can work with Liqo with no modifications.

Reviewer: ```suggestion
This is a traditional [Kubernetes Service](https://kubernetes.io/docs/concepts/services-networking/service/), which can work with Liqo without any modification.
```

Author: This section is removed following Marco Iorio's suggestions.

> Indeed, inspecting the Service, it is possible to observe that both `nginx` pods are correctly specified as endpoints.

Reviewer: ```suggestion
Indeed, inspecting the Service, it shows that both `nginx` pods are correctly specified as endpoints.
```

Author: This section is removed following Marco Iorio's suggestions.

```bash
kubectl describe service liqo-demo -n liqo-demo
```

```text
Name:              liqo-demo
Namespace:         liqo-demo
Labels:            <none>
Annotations:       <none>
Selector:          app=liqo-demo
Type:              ClusterIP
IP Family Policy:  SingleStack
IP Families:       IPv4
IP:                10.93.84.150
IPs:               10.93.84.150
Port:              web  80/TCP
TargetPort:        web/TCP
Endpoints:         10.200.1.10:80,10.202.1.10:80
Session Affinity:  None
Events:            <none>
```

#### Check the Service connectivity

It is now possible to contact the Service: as usual, Kubernetes will forward the HTTP request to one of the available back-end pods.
Additionally, all traditional mechanisms still work seamlessly (e.g. DNS discovery), even though one of the pods is actually running in a _remote_ cluster.

You can fire up a pod and run `curl` from inside the cluster:

```bash
kubectl run --image=curlimages/curl curl -n default -it --rm --restart=Never -- \
    curl --silent liqo-demo.liqo-demo.svc.cluster.local | grep 'Server'
```

```{admonition} Note

> Executing the previous command multiple times, you will observe that part of the requests are answered by the pod running in the *local* cluster, and in part by that in the *remote* cluster (i.e., the Server value changes).

Reviewer: ```suggestion
Executing the previous command multiple times, you will observe that part of the requests are answered by the pod running in the *local* cluster, and in part by that in the *remote* cluster (i.e., the `Server` value changes).
```

Author: This section is removed following Marco Iorio's suggestions.

```

## Tear down the infrastructure

To tear down all the infrastrucure you only need to run the following command:
```bash
terraform destroy -auto-approve
```

This command will destroy all the resource starting from the last one created, from bottom to top.\
If you want to destroy a specific resource (for example to unpeer or to unoffload) you can leverage on the `-target` flag of `destroy` command.

> For example to unpeer two cluster you can run the following command:

Reviewer: ```suggestion
For example, you can run the following command to un-peer two clusters:
```

Author: Committed correctly.
